chips/swerv/src/eh1_timer.rs

```diff
@@ -0,0 +1,213 @@
+//! Internal Timer
+
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::register_bitfields;
+use kernel::hil::time;
+use kernel::hil::time::{Alarm, Counter, Ticks, Ticks32, Time};
+use kernel::ReturnCode;
+use riscv_csr::csr::ReadWriteRiscvCsr;
+
+/// 100Hz `Frequency`
+#[derive(Debug)]
+pub struct Freq100Hz;
+impl time::Frequency for Freq100Hz {
+    fn frequency() -> u32 {
+        100
+    }
+}
+
+pub enum TimerNumber {
+    ZERO,
+    ONE,
+}
+
+register_bitfields![usize,
+    MITCNT [
+        COUNT OFFSET(0) NUMBITS(32) []
+    ],
+    MITB [
+        BOUND OFFSET(0) NUMBITS(32) []
+    ],
+    MITCTL [
+        ENABLE OFFSET(0) NUMBITS(1) [],
+        HALT_EN OFFSET(1) NUMBITS(1) [],
+        PAUSE_EN OFFSET(2) NUMBITS(1) [],
+    ],
+];
+
+#[allow(dead_code)]
+pub struct Timer<'a> {
+    number: TimerNumber,
+    alarm_client: OptionalCell<&'a dyn time::AlarmClient>,
+
+    mitcnt0: ReadWriteRiscvCsr<usize, MITCNT::Register, 0x7D2>,
+    mitcnt1: ReadWriteRiscvCsr<usize, MITCNT::Register, 0x7D5>,
+
+    mitb0: ReadWriteRiscvCsr<usize, MITB::Register, 0x7D3>,
+    mitb1: ReadWriteRiscvCsr<usize, MITB::Register, 0x7D6>,
+
+    mitctl0: ReadWriteRiscvCsr<usize, MITCTL::Register, 0x7D4>,
+    mitctl1: ReadWriteRiscvCsr<usize, MITCTL::Register, 0x7D7>,
+}
+
+impl Timer<'_> {
+    pub const fn new(number: TimerNumber) -> Self {
+        Timer {
+            number,
+            alarm_client: OptionalCell::empty(),
+            mitcnt0: ReadWriteRiscvCsr::new(),
+            mitcnt1: ReadWriteRiscvCsr::new(),
+            mitb0: ReadWriteRiscvCsr::new(),
+            mitb1: ReadWriteRiscvCsr::new(),
+            mitctl0: ReadWriteRiscvCsr::new(),
+            mitctl1: ReadWriteRiscvCsr::new(),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        self.stop();
```

Comment:
Why stop the timer here? Could it not just be allowed to continue?

Reply:
It can, but then you will get another interrupt when it ticks over the compare register again. The hardware automatically rests the counter when an interrupt occurs.

Reply:
oh, makes sense.